### PR TITLE
Optimize GROUP BY on single string column

### PR DESCRIPTION
**Pull request by mfussenegger**
This adds a new specialized BatchIterator implementation which can be
used if a GROUP BY on shard-level for single string column is done.

This specialized BatchIterator implementation utilizes Lucene ordinal
values to reduce the number of key value retrievals from one per row to
one per unique key value.

Cases which should see a big improvement are GROUP BY on low cardinality
columns with few aggregations.

```
V1: 3.1.0-c3a321f28a9a0abd1d963180007f88aeaaee13be
V2: 3.1.0-8e7f59712ca26de466832694da2cfb2ea3109c55

Q: select "cCode", count(*) from uservisits group by "cCode"
C: 1
  147.71% mean difference. Likely significant
             V1    →   V2
  mean:  339.379 →  51.034
  stdev:  28.136 →  38.005
  max:   589.710 → 422.792
  min:   313.943 →  34.308

Q: select min("adRevenue") from uservisits group by "cCode"
C: 1
  68.32% mean difference. Likely significant
             V1    →   V2
  mean:  364.508 → 178.885
  stdev:  14.848 →  75.244
  max:   409.392 → 789.264
  min:   334.004 → 138.263

Q: select max("adRevenue") from uservisits group by "cCode"
C: 1
  76.70% mean difference. Likely significant
             V1    →   V2
  mean:  365.177 → 162.723
  stdev:  16.056 →  45.588
  max:   447.554 → 590.674
  min:   339.256 → 135.773

Q: select avg("adRevenue") from uservisits group by "cCode"
C: 1
  79.51% mean difference. Likely significant
             V1    →   V2
  mean:  363.576 → 156.731
  stdev:  14.117 →  10.758
  max:   403.539 → 190.785
  min:   334.786 → 129.357

Q: select count(distinct "searchWord") from uservisits group by "cCode"
C: 1
  33.78% mean difference. Likely significant
             V1    →   V2
  mean:  1046.754 → 744.235
  stdev:  39.760 →  68.410
  max:   1305.067 → 1312.638
  min:   980.169 → 685.295

Q: select arbitrary("searchWord") from uservisits group by "cCode"
C: 1
  58.21% mean difference. Likely significant
             V1    →   V2
  mean:  722.474 → 396.746
  stdev:  77.180 →  12.042
  max:   1179.716 → 427.690
  min:   661.673 → 370.212
```

### CODE REVIEW

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java [hunk 1]
```diff
@@ -0,0 +1,357 @@
...
+                                       LuceneQueryBuilder luceneQueryBuilder,
+                                       FieldTypeLookup fieldTypeLookup,
+                                       BigArrays bigArrays,
+                                       InputFactory inputFactory,
+                                       DocInputFactory docInputFactory,
+                                       RoutedCollectPhase collectPhase,
+                                       CollectTask collectTask) {
+        Collection<? extends Projection> shardProjections = shardProjections(collectPhase.projections());
+        GroupProjection groupProjection = getSingleStringKeyGroupProjection(shardProjections);
+        if (groupProjection == null) return null;
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: can we use braces please?

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java [hunk 2]
```diff
@@ -0,0 +1,357 @@
...
+
+        assert groupProjection.keys().size() == 1 : "Must have 1 key if getSingleStringKeyGroupProjection returned a projection";
+        Reference keyRef = getKeyRef(collectPhase.toCollect(), groupProjection.keys().get(0));
+        if (keyRef == null) {
+            return null; // group by on non-reference
+        }
+        MappedFieldType keyFieldType = fieldTypeLookup.get(keyRef.column().fqn());
+        if (keyFieldType == null || !keyFieldType.hasDocValues()) {
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: hmm why are the doc values a deal breaker here?

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java [hunk 3]
```diff
@@ -0,0 +1,357 @@
...
+                        //noinspection unchecked
+                        cells[c] = mode.finishCollect(ramAccounting, aggregations.get(i).function(), states[i]);
+                    }
+                    return row;
+                }
+            })
+            .iterator();
+    }
+
+    private static Map<BytesRef, Object[]> getGroupedStates(BigArrays bigArrays,
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: can we document this? (or rename it to something that indicates what kind of grouping it's performing - groupedStatesByOrdinalRef maybe? - or something along those lines)

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java [hunk 4]
```diff
@@ -0,0 +1,357 @@
...
+                }
+                for (long ord = 0; ord < statesByOrd.size(); ord++) {
+                    Object[] states = statesByOrd.get(ord);
+                    if (states == null) {
+                        continue;
+                    }
+                    BytesRef sharedKey = values.lookupOrd(ord);
+                    Object[] prevStates = statesByKey.get(sharedKey);
+                    if (prevStates == null) {
+                        long hashMapEntryOverhead = 36L;
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: how's this magic number come up? maybe document that?

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java [hunk 5]
```diff
@@ -0,0 +1,357 @@
...
+            //noinspection unchecked
+            states[i] = aggregation.function().iterate(
+                ramAccounting,
+                states[i],
+                aggregation.inputs()
+            );
+        }
+    }
+
+    private static Object[] initStates(BigArrays bigArrays,
```
Comment (andreidan) [inline sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: can we be more exact on what "states" we actually initiate? this is super generic and without browsing the method body is not exactly clear what's doing

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java [hunk 6]
```diff
@@ -0,0 +1,357 @@
...
+                            BytesRefSizeEstimator.INSTANCE.estimateSize(sharedKey) + hashMapEntryOverhead));
+                        statesByKey.put(BytesRef.deepCopyOf(sharedKey), states);
+                    } else {
+                        for (int i = 0; i < aggregations.size(); i++) {
+                            AggregationContext aggregation = aggregations.get(i);
+                            //noinspection unchecked
+                            prevStates[i] = aggregation.function().reduce(
+                                ramAccounting,
+                                prevStates[i],
+                                states[i]
```
[the remark this answers is not in the document]
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: `aggregateValues` is using `function().iterate(..)`, here I'm using `function.reduce()`, the signatures of these methods are also different so passing in a lambda is probably overkill and more complicated than having it like this.

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java (hunk 2 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: The ordinal values wouldn't be available without doc-values.

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java (hunk 5 above)
Comment (mfussenegger) [reply sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: I'm not sure. Any suggestions? I think `initAggregationStates` doesn't make it more readable as there are no other states in the scope of this class. Maybe we need to refer to the states of the aggregations with a different name, but I think it's quite fitting and clear.

File: sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java (hunk 6 above)
[the remark this answers is not in the document]
Comment (seut) [reply sql/src/main/java/io/crate/execution/engine/collect/GroupByOptimizedIterator.java]: ah right, didn't read carefully enough ;)

**Pull request merged**